rest/main.go

```diff
@@ -146,3 +158,144 @@ func serverMainPersistentConfig(fs *flag.FlagSet, flagStartupConfig *StartupConf
 
 	return false, startServer(&sc, ctx)
 }
+
+func automaticConfigUpgrade(configPath string) (*StartupConfig, bool, error) {
+	legacyServerConfig, err := LoadServerConfig(configPath)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if legacyServerConfig.DisablePersistentConfig != nil && *legacyServerConfig.DisablePersistentConfig {
+		return nil, true, nil
+	}
+
+	startupConfig, dbConfigs, err := legacyServerConfig.ToStartupConfig()
+	if err != nil {
+		return nil, false, err
+	}
+
+	dbConfigs, err = sanitizeDbConfigs(dbConfigs)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Attempt to establish connection to server, add retry like its other use-case
+	cluster, err := EstablishCouchbaseClusterConnection(startupConfig)
+	if err != nil {
+		return nil, false, err
+	}
+
+	defer func() {
+		_ = cluster.Close()
+	}()
+
+	// Write database configs to CBS with groupID "default"
+	for _, dbConfig := range dbConfigs {
+		_, err = cluster.PutConfig(*dbConfig.Bucket, "default", base.Uint64Ptr(0), dbConfig)
+		if err != nil {
+			// TODO: if exists skip, else error
+			return nil, false, err
+		}
+	}
+
+	// Attempt to backup current config
+	err = backupCurrentConfigFile(configPath)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Overwrite old config with new migrated startup config
+	jsonStartupConfig, err := json.Marshal(startupConfig)
+	err = ioutil.WriteFile(configPath, jsonStartupConfig, 0644)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return startupConfig, false, nil
+}
+
+// validate / sanitize db configs
+// - remove servers
+// - remove users
+// - ensure servers are the same
+func sanitizeDbConfigs(configMap DbConfigMap) (DbConfigMap, error) {
+	var databaseServerAddress string
+
+	for dbName, dbConfig := range configMap {
+		if databaseServerAddress == "" {
+			databaseServerAddress = *dbConfig.Server
+		}
+
+		if *dbConfig.Server != databaseServerAddress {
+			return nil, fmt.Errorf("server addresses specified in dbConfig do not match. This is required for " +
+				"persistent config")
+		}
+
+		if dbConfig.Bucket == nil || *dbConfig.Bucket == "" {
+			*dbConfig.Bucket = dbName
```

Comment:
I think you probably want this? (and copy to avoid the variable changing during iteration)

```suggestion
			dbNameCopy := dbName
			dbConfig.Bucket = &dbNameCopy
```

Reply:
Yikes... yeah.